nest.js

```
@@ -472,6 +473,66 @@ class NestThermostat extends NestDevice {
 			}).catch(err => console.error(err));
 		});
 	}
+
+	/**
+	 * Check if devce software version is greater than or
+	 * equal to the provided version parameter.
+	 * @param version
+	 * @returns {boolean}
+	 */
+	checkSoftwareVersionGTE(version) {
+		let major = this.software_version.split(".")[0];
+		let minor = this.software_version.split(".")[1];
+		return (major >= version.split(".")[0] && minor >= version.split(".")[1]);
+	}
+
+	/**
+	 * Set the target HVAC mode of this Nest Thermostat.
+	 * @param mode
+	 */
+	setHvacMode(mode) {
+		return new Promise((resolve, reject) => {
+
+			// Authenticate
+			this.nest_account.authenticate().then(() => {
+
+				// Handle cases where mode is unsupported
+				if (this.is_using_emergency_heat) {
+					return reject(__('error.hvac_emergency_heat', {
+						name: this.name_long,
+					}));
+				} else if (mode === 'heat-cool' && !(this.can_cool && this.can_heat)) {
+					return reject(__('error.hvac_mode_heat-cool_unsupported', {
+						name: this.name_long,
+					}));
+				} else if (mode === 'cool' && !this.can_cool) {
+					return reject(__('error.hvac_mode_cool_unsupported', {
+						name: this.name_long,
+					}));
+				} else if (mode === 'heat' && !this.can_heat) {
+					return reject(__('error.hvac_mode_heat_unsupported', {
+						name: this.name_long,
+					}));
+				} else if (mode === 'eco' && (!this.checkSoftwareVersionGTE("5.6.0") || !(this.can_cool || this.can_heat))) {
+					return reject(__('error.hvac_mode_eco_unsupported', {
+						name: this.name_long,
+					}));
+				}
+
+				// All clear to change the HVAC mode
+				this.nest_account.db.child(`devices/thermostats/${this.device_id}/hvac_mode`).set(mode, error => {
+					if (error) {
+						return reject(__('error.unknown', {
+							hvac_mode: mode,
+							name: this.name_long,
+							error,
+						}));
+					}
+					return resolve(mode);
+				});
+			}).catch(err => console.error(err));
```

this catch should re-throw the error so callers of `setHvacMode` also get the error in the catch clause